Fix deallocate of unallocated pointers

underwoo opened the pull request:
**Description**
This commit fixes the test_fms/fms tests so they now run correctly.  This required fixes to two calls to fms_io_unstructured_register_restart_axis and reworks deallocate_unstruct_pass_type to correcly deallocate and nullify pointers

Fixes #649

**How Has This Been Tested?**
Ran tests on Mac OS X 10.12 and through the CI process

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [ ] `make distcheck` passes

underwoo committed:
Fix deallocate of unallocated pointers

This commit fixes the test_fms/fms tests so they now run correctly.  This required fixes to two calls to fms_io_unstructured_register_restart_axis and reworks deallocate_unstruct_pass_type to correcly deallocate and nullify pointers

underwoo commented:
The tests ran, and appears as if the two additional (no longer skipped) tests are running and passing.  @menzel-gfdl please review the changes, and let us know if these are acceptable so we can merge this in and have a few more tests.

menzel-gfdl commented:
@underwoo The changes look good to me.  Thanks.

bensonr reviewed (commented):
In deallocate_unstruct_pass_type, you only check for UG2SG association, and use that to nullify the SG2UG entry.  Why are we not testing for and then deallocating the send/recv members of the SG2UG struct as well for completeness sake?

underwoo commented:
> In deallocate_unstruct_pass_type, you only check for UG2SG association, and use that to nullify the SG2UG entry. Why are we not testing for and then deallocating the send/recv members of the SG2UG struct as well for completeness sake?

In reviewing the code for this fix, I found that `UG2SG%send` points to `SG2UG%recv` and `UG2SG%recv` points to `SG2UG%send`.  The original code did check if both were associated, and then attempted to deallocate the members of `SG2UG`.  The pointers `%send` and `%recv` were still associated, but the memory they pointed to was deallocated when the `UG2SG` members were deallocate.  That is why they are simply `nullify`ed instead of checking if allocated. 

I'll add additional documentation to the code to make that clear.  Something I should have done originally.  If you wouldn't mind, please review and click request changes.

underwoo committed:
Add comments to explain why only deallocate UG2SG%{send,recv}

Added comments to explain why the code only deallocate `UG2SG%{send,recv}` and nullifies `SG2UG%{recv,send}` as they both point to the same memory.

underwoo commented:
I have updated the comments as mentioned in the previous comment.  Please re-review.
